### DISCUSSION TITLE
Express Session not persisting

### DISCUSSION THREAD

**Discussion by claidler** [trpc/trpc#2570] [Q&A] [answered]
Are there any examples using session cookies, React and express and a  login route. This was working for me before I added trpc, so I'm obviously doing something wrong, but not sure what? Any help would be much appreciated as I'm desperate to use it! 

server/context

```typescript
import * as trpc from '@trpc/server';
import * as trpcExpress from '@trpc/server/adapters/express';

export async function createContext(
	opts?: trpcExpress.CreateExpressContextOptions
) {
	async function getSessionData() {
		if (opts?.req.session) {
			return { ...opts.req.session };
		}
		return null;
	}
	const sessionData = await getSessionData();
	return {
		...sessionData
	};
}
export type Context = trpc.inferAsyncReturnType<typeof createContext>;

export function createRouter() {
	return trpc.router<Context>();
}
```

/server/routes

```typescript
export const contactRouter = createRouter()
	.mutation('login', {
		input: z.object({
			username: z.string(),
			password: z.string().nullable(),
		}),
		resolve: async ({ input, ctx }) => {

// check login details in db and return contact from db

					ctx.userId = returnedContactFromDb.id
				});
			}
		}
       )
```

Now ctx.userId doesn't seem to be accessible on each request. 
I don't think it's a CORS issue.

on the client I have:

```typescript
	const [trpcClient] = useState(() =>
		trpc.createClient({
			url: 'http://localhost:5000',
			fetch(url, options) {
				return fetch(url, {
					...options,
					credentials: 'include',
					mode: 'cors',
				});
			},
		})
	);
```

**Comment by OrJDev** [comment, marked as answer]
i could help you if you shared the link to the full repo, this code doesn't really say much 

Edit: and what do you mean by not accessible? Like typescript throws an error that userId is not a part of context or its not being saved at all?

**Reply by claidler** [reply]
Thanks @OrJDev unfortunately I can't share the code here, which is why I only shared these snippets.
Yeah 'not accessible' probably wasn't the best wording there. I just meant it's undefined. I can access the userId on the context, but it's not persisting in the session.

**Reply by claidler** [reply]
I just mostly wondered if there was an example for just a standard express server without nextjs and nextjsauth or jwt's being involved. Really basic express API example with no SSR or JWT's or anything. Just session cookies and express

**Reply by OrJDev** [reply]
Alright i will do that today and will send you the repo.

**Reply by OrJDev** [reply]
![Screenshot_1](https://user-images.githubusercontent.com/91349014/188297526-92f97980-9c76-462c-8430-073a59de0262.png)

https://github.com/OrJDev/trpc-express-session

i hope thats what you meant, if its not then let me know and i will try to make it.

EDIT: also don't forget to add this line when you create new trpc client:

```javascript
  fetch(url, options) {
    return fetch(url, {
      ...options,
      credentials: "include",
    });
  },
```

**Reply by claidler** [reply]
Thanks so much! I will have another go and see if I can figure out where I was going wrong

**Reply by claidler** [reply]
@OrJDev sorry to waste your time 😑 I just realised the problem... I am using trpc in a monorepo, all apps use the same login page, and I was running the wrong app - one where I hadn't setup trpc client side. Works great. I looked at your code and looked at mine and couldn't see what the problem with mine could be comparibly! It was helpful for that reason, forced me to relook at mine. Also yours is cleaner than mine in places (context in particular) so will do some changes there! Really appreciate the help

**Reply by OrJDev** [reply]
No problem i was getting ready for school and was bored anyways, happy i could help.

**Reply by brian-chano** [reply]
Hey @OrJDev and @claidler 

I happen to have a similar problem, I think.

I can't get my session to persist. Using a monorepo too, but I have double-checked my client and it's configured correctly.
My request session is always undefined.
The only difference I can tell is I'm trying to use [websockets](https://trpc.io/docs/v10/subscriptions), so I end up typing my context parameters differently.

I have this in my types definition to get my web subscription to work with my [session ](https://socket.io/how-to/use-with-express-session#with-typescript)

```typescript
import { Session } from 'express-session'

declare module 'http' {
  interface IncomingMessage {
    session: Session & {
      userId?: string
    }
  }
}
```
Here is my server
```typescript
// Express
const app = express()
const server = createServer(app)
// Subscription server
const wss = new Server({ server })
const handler = applyWSSHandler<AppRouter>({
  wss,
  router,
  createContext,
})

// express-session
app.use(session({
  name: "example",
  cookie: {
    http: process.env.NODE_ENV === "production",
    samesite: "lax",
    secure: process.env.NODE_ENV === "production",
  }
}))

app.use(
  '/api',
  createExpressMiddleware({
    router,
    createContext,
  })
)
```
Here is my context for example 

```typescript
export const createContext = async ({
  req,
  res,
}: CreateExpressContextOptions | CreateWSSContextFnOptions) => {

// console.log(req.session) 
// is undefined here
  return {
    req,
    res,
    prisma,
  }
}

export type Context = trpc.inferAsyncReturnType<typeof createContext>


If I remove all the web socket stuff, everything works fine.

I tried to access the repo you linked, but it's sadly gone. Would appreciate if you could point me to it, so I can compare some notes.

**Reply by OrJDev** [reply]
> Hey @OrJDev and @claidler 
> 
> 
> 
> I happen to have a similar problem, I think.
> 
> 
> 
> I can't get my session to persist. Using a monorepo too, but I have double-checked my client and it's configured correctly.
> 
> My request session is always undefined.
> 
> The only difference I can tell is I'm trying to use [websockets](https://trpc.io/docs/v10/subscriptions), so I end up typing my context parameters differently.
> 
> 
> 
> I have this in my types definition to get my web subscription to work with my [session ](https://socket.io/how-to/use-with-express-session#with-typescript)
> 
> 
> 
> ```typescript
> 
> import { Session } from 'express-session'
> 
> 
> 
> declare module 'http' {
> 
>   interface IncomingMessage {
> 
>     session: Session & {
> 
>       userId?: string
> 
>     }
> 
>   }
> 
> }
> 
> ```
> 
> Here is my server
> 
> ```typescript
> 
> // Express
> 
> const app = express()
> 
> const server = createServer(app)
> 
> // Subscription server
> 
> const wss = new Server({ server })
> 
> const handler = applyWSSHandler<AppRouter>({
> 
>   wss,
> 
>   router,
> 
>   createContext,
> 
> })
> 
> 
> 
> // express-session
> 
> app.use(session({
> 
>   name: "example",
> 
>   cookie: {
> 
>     http: process.env.NODE_ENV === "production",
> 
>     samesite: "lax",
> 
>     secure: process.env.NODE_ENV === "production",
> 
>   }
> 
> }))
> 
> 
> 
> app.use(
> 
>   '/api',
> 
>   createExpressMiddleware({
> 
>     router,
> 
>     createContext,
> 
>   })
> 
> )
> 
> ```
> 
> Here is my context for example 
> 
> 
> 
> ```typescript
> 
> export const createContext = async ({
> 
>   req,
> 
>   res,
> 
> }: CreateExpressContextOptions | CreateWSSContextFnOptions) => {
> 
> 
> 
> // console.log(req.session) 
> 
> // is undefined here
> 
>   return {
> 
>     req,
> 
>     res,
> 
>     prisma,
> 
>   }
> 
> }
> 
> 
> 
> export type Context = trpc.inferAsyncReturnType<typeof createContext>
> 
> 
> 
> 
> 
> If I remove all the web socket stuff, everything works fine.
> 
> 
> 
> I tried to access the repo you linked, but it's sadly gone. Would appreciate if you could point me to it, so I can compare some notes. 

restored the repo

https://github.com/OrJDev/trpc-express-session

Hope it helps

**Reply by brian-chano** [reply]
> > Hey @OrJDev and @claidler
> > I happen to have a similar problem, I think.
> > I can't get my session to persist. Using a monorepo too, but I have double-checked my client and it's configured correctly.
> > My request session is always undefined.
> > The only difference I can tell is I'm trying to use [websockets](https://trpc.io/docs/v10/subscriptions), so I end up typing my context parameters differently.
> > I have this in my types definition to get my web subscription to work with my [session ](https://socket.io/how-to/use-with-express-session#with-typescript)
> > ```ts
> > import { Session } from 'express-session'
> > 
> > 
> > 
> > declare module 'http' {
> > 
> >   interface IncomingMessage {
> > 
> >     session: Session & {
> > 
> >       userId?: string
> > 
> >     }
> > 
> >   }
> > 
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Here is my server
> > ```ts
> > // Express
> > 
> > const app = express()
> > 
> > const server = createServer(app)
> > 
> > // Subscription server
> > 
> > const wss = new Server({ server })
> > 
> > const handler = applyWSSHandler<AppRouter>({
> > 
> >   wss,
> > 
> >   router,
> > 
> >   createContext,
> > 
> > })
> > 
> > 
> > 
> > // express-session
> > 
> > app.use(session({
> > 
> >   name: "example",
> > 
> >   cookie: {
> > 
> >     http: process.env.NODE_ENV === "production",
> > 
> >     samesite: "lax",
> > 
> >     secure: process.env.NODE_ENV === "production",
> > 
> >   }
> > 
> > }))
> > 
> > 
> > 
> > app.use(
> > 
> >   '/api',
> > 
> >   createExpressMiddleware({
> > 
> >     router,
> > 
> >     createContext,
> > 
> >   })
> > 
> > )
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Here is my context for example
> > ```ts
> > export const createContext = async ({
> > 
> >   req,
> > 
> >   res,
> > 
> > }: CreateExpressContextOptions | CreateWSSContextFnOptions) => {
> > 
> > 
> > 
> > // console.log(req.session) 
> > 
> > // is undefined here
> > 
> >   return {
> > 
> >     req,
> > 
> >     res,
> > 
> >     prisma,
> > 
> >   }
> > 
> > }
> > 
> > 
> > 
> > export type Context = trpc.inferAsyncReturnType<typeof createContext>
> > 
> > 
> > 
> > 
> > 
> > If I remove all the web socket stuff, everything works fine.
> > 
> > 
> > 
> > I tried to access the repo you linked, but it's sadly gone. Would appreciate if you could point me to it, so I can compare some notes. 
> > ```
> 
> restored the repo
> 
> https://github.com/OrJDev/trpc-express-session
> 
> Hope it helps

Thanks a bunch. Let me take a look :)

**Reply by OrJDev** [reply]
Also make sure you didn't forget to add this line when creating new trpc client.

```ts
   fetch(url, options) {
    return fetch(url, {
      ...options,
      credentials: "include",
    });
  },
```

**Reply by brian-chano** [reply]
Yeah! i already have this in my client. But I'm trying to use websockets. Will play around with your repo and see what I get
